File: src/models/selection/RouletteSelection.ts
import ISelection from "./ISelection";
import Population from "../population/Population";
import Individual from "../Individual";

export default class RouletteSelection implements ISelection {
    evaluate(population: Population): Individual {
        let maxLength = 0,
            minLength = Number.MAX_VALUE;

        population.individuals.forEach(individual => {
            const distance = individual.calculateTotalDistance();

            if (distance > maxLength) maxLength = distance;
            if (distance < minLength) minLength = distance;
        });

        let sumFitness = 0;
        population.individuals.forEach(
            individual =>
                (sumFitness +=
                    (minLength * 1) / individual.calculateTotalDistance())
        );

        let sum = 0;

        population.individuals.forEach(individual => {
            sum += (minLength * 1) / individual.calculateTotalDistance();

            if (Math.random() < sum) return individual;
        });

        throw new Error("Eh...");
    }
}

File: zad1/Algorithms/Selection/RouletteSelection.cs
using System;
using System.Linq;
using zad1.Models;

namespace zad1.Algorithms.Selection
{
    public class RouletteSelection : ISelection
    {
        public Individual Evaluate(Population population)
        {
            var sumFitness =
                population.Individuals.Sum(individual => 1 / Math.Pow(individual.CalculateTotalDistance(), 10));

            var rng = new Random();
            var random = rng.NextDouble() * sumFitness;

            double sum = 0;
            foreach (var individual in population.Individuals)
            {
                sum += 1 / Math.Pow(individual.CalculateTotalDistance(), 10);

                if (sum >= random) return individual;
            }

            return null;
        }
    }
}